pkg/p2p/kadcli/peer_test.go

```diff
@@ -0,0 +1,289 @@
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"testing"
+
+	"google.golang.org/grpc"
+
+	"github.com/dusk-network/dusk-blockchain/pkg/core/data/block"
+	"github.com/dusk-network/dusk-blockchain/pkg/core/tests/helper"
+	"github.com/dusk-network/dusk-blockchain/pkg/p2p/peer"
+	"github.com/dusk-network/dusk-blockchain/pkg/p2p/wire/checksum"
+	"github.com/dusk-network/dusk-blockchain/pkg/p2p/wire/message"
+	"github.com/dusk-network/dusk-blockchain/pkg/p2p/wire/protocol"
+	"github.com/dusk-network/dusk-blockchain/pkg/p2p/wire/topics"
+	"github.com/dusk-network/dusk-blockchain/pkg/rpc/client"
+	"github.com/dusk-network/dusk-blockchain/pkg/util/nativeutils/eventbus"
+	"github.com/dusk-network/dusk-protobuf/autogen/go/rusk"
+	"github.com/stretchr/testify/assert"
+)
+
+const (
+	RUSK_ADDR = "127.0.0.1:50005"
+)
+
+// TestListenStreamReader tests the kadcli.Reader receiving a block
+// from the rusk:Network.Listen stream-based endpoint.
+func TestListenStreamReader(t *testing.T) {
+	assert := assert.New(t)
+	rcvChan := make(chan message.Message)
+	errChan := make(chan error)
+
+	// basic infrastructure
+	eb := eventbus.New()
+	p := peer.NewMessageProcessor(eb)
+	g := protocol.NewGossip(protocol.TestNet)
+
+	// have a callback ready for Block topic
+	respFn := func(srcPeerID string, m message.Message) ([]bytes.Buffer, error) {
+		if srcPeerID != RUSK_ADDR {
+			errChan <- errors.New("callback: wrong source address")
+			return nil, nil
+		}
+		rcvChan <- m
+		return nil, nil
+	}
+	p.Register(topics.Block, respFn)
+
+	// create a rusk mock server
+	srv, err := NewRuskMock(g, errChan)
+	if err != nil {
+		t.Fatalf("rusk mock failed: %v", err)
+	}
+
+	// create a client that connects to rusk mock server
+	_, grpcConn := client.CreateNetworkClient(context.Background(), RUSK_ADDR)
+	ruskc := rusk.NewNetworkClient(grpcConn)
+
+	// create our kadcli (gRPC) Reader
+	r := NewReader(eb, g, p, ruskc)
+
+	// subscribe to gRPC stream
+	go r.Listen()
+
+	// process status/output
+	select {
+	case err := <-errChan:
+		t.Fatal(err)
+	case m := <-rcvChan:
+		b := m.Payload().(block.Block)
+		assert.True(m.Category() == topics.Block)
+		assert.True(b.Header.Height == 5525)
+		assert.True(len(b.Txs) == 12)
+	}
+
+	// Disconnect client and stop server
+	grpcConn.Close()
+	srv.Stop()
+}
+
+// TestBroadcastWriter tests the kadcli.Writer by broadcasting
+// a block message through a mocked rusk client.
+func TestBroadcastWriter(t *testing.T) {
+	assert := assert.New(t)
+	rcvChan := make(chan *rusk.BroadcastMessage)
+
+	// Basic infrastructure
+	eb := eventbus.New()
+	g := protocol.NewGossip(protocol.TestNet)
+
+	// create a mock client
+	cli := NewMockNetworkClient(rcvChan)
+
+	// create our kadcli Writer
+	w := NewWriter(eb, cli)
+
+	// create a mock message
+	buf, err := createBlockMessage()
+	if err != nil {
+		t.Errorf("fail to create msg: %v", err)
+	}
+
+	go func() {
+		// send a broadcast message
+		if errw := w.WriteToAll(buf.Bytes(), []byte{127}, 0); errw != nil {
```

Comment:
We should use `eventbus.Publish(topics.Kadcast, ...`  so that we can ensure both subscription and message handling are working properly.